src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -1061,6 +1077,22 @@ static int client_send_discover(sd_dhcp_client *client) {
         if (r < 0)
                 return r;
 
+        /* RFC1542 section 3.5:
+           if the client has no information to communicate to the server,
+           the octet immediately following the magic cookie SHOULD be set
+           to the "End" tag (255) and the remaining octets of the 'vend'
+           field SHOULD be set to zero.
+         */
+        /* Use this RFC, along with the fact that some BOOTP servers require
+           a 64-byte vend field, to suggest that we always zero and send 64
+           bytes in the options field.
+         */
+        if (client->bootp)
+                if (optoffset < 64 && optlen >= 64) {
+                        memset(&discover->dhcp.options[optoffset], 0, optlen - optoffset);
+                        optoffset = 64;
```

Comment:
I took a quick look at BOOTP requests networkd sends and as far as I can see the length of the vend field networkd sends is 68 instead of 64 (https://www.rfc-editor.org/rfc/rfc951#section-3). I think it's because `DHCPMessage` wasn't designed with BOOTP in mind so it includes `magic` (4 bytes) and `options[]` (with its variable length). To turn it all into the vend field compliant with the BOOTP RFC the length of `options[]` should be 60.

As far as I can see in bootp mode networkd still sends a bunch of DHCP options
```sh
    Magic cookie: DHCP                                                                                                                        00:33:02 [6/384]
    Option: (55) Parameter Request List                                                                                                                       
        Length: 12                                                                                                                                            
        Parameter Request List Item: (1) Subnet Mask                                                                                                          
        Parameter Request List Item: (3) Router                                                                                                               
        Parameter Request List Item: (6) Domain Name Server                                                                                                   
        Parameter Request List Item: (12) Host Name                                                                                                           
        Parameter Request List Item: (15) Domain Name
        Parameter Request List Item: (33) Static Route
        Parameter Request List Item: (42) Network Time Protocol Servers
        Parameter Request List Item: (101) TCode
        Parameter Request List Item: (114) DHCP Captive-Portal
        Parameter Request List Item: (120) SIP Servers
        Parameter Request List Item: (121) Classless Static Route
        Parameter Request List Item: (162) Unassigned
    Option: (57) Maximum DHCP Message Size
        Length: 2
        Maximum DHCP Message Size: 1472 
    Option: (80) Rapid commit
        Length: 0
        Value: <MISSING>
    Option: (12) Host Name
        Length: 7
        Length: 7
        Host Name: default
    Option: (255) End
        Option End: 255
    Padding: 00000000000000000000000000000000000000000000000000000000000000000000
```
and I don't think it should send any of them.

(I can confirm that when networkd is stopped no BOOTP requests are sent)

Reply:
In theory, it's not important to avoid sending DHCP options to a BOOTP server, it should ignore any options it doesn't understand. That's what the magic cookie is for, and that was already present in RFC951. Granted, the `99.130.83.99` value for generic extensions hadn't been established yet. Perhaps it's sufficient to "hope" that nobody who implemented their own vendor-specific extensions before RFC1048 happened to choose that exact value for themselves.

Reply:
I think in theory it shouldn't be necessary to omit the DHCP Message Type option either (because plain BOOTP servers should ignore it too) but apparently there are BOOTP servers out there where it's recognized. I'd err on the side of being as close to `bootpc` as possible by default to avoid any clashes with whatever can be rolled out "in the wild". I think it should be OK to allow appending more options at some point but it can probably be done later if anyone is interested in it.

Before I forget @avramd could you show what BOOTP replies look like in your environment and attach it to https://github.com/systemd/systemd/issues/34885?

Reply:
>I'd err on the side of being as close to `bootpc` as possible by default to avoid any clashes with whatever can be rolled out "in the wild". I think it should be OK to allow appending more options at some point but it can probably be done later if anyone is interested in it.

@evverx I've explicitly commented out all options and the request looks the same as `bootpc` now. Only the END followed by 0-padding to the 64 byte limit. I'll look for any more pick-ups required.

@avramd I'm adding your bootp reply packet to `test-dhcp-client.c` and will use that to verify both of our use cases. I hope to push a new version today.

General question: should I rebase on main before the push? I was looking at my push history and I see it might be difficult to track versions if I keep doing that.